pkl-core/src/main/java/org/pkl/core/project/Project.java

```diff
@@ -126,11 +124,97 @@ public static Project load(Evaluator evaluator, ModuleSource moduleSource) {
     try {
       var output = evaluator.evaluateOutputValueAs(moduleSource, PClassInfo.Project);
       return Project.parseProject(output);
+    } catch (StackOverflowError e) {
+      var cycles = findImportCycle(moduleSource);
+      if (!cycles.isEmpty()) {
+        var cycle = cycles.stream().map(Project::renderCycle).collect(Collectors.joining("\n"));
+        throw new VmExceptionBuilder()
+            .evalError("cannotHaveCircularProjectDependencies", cycle)
+            .withCause(e)
+            .build()
+            .toPklException(evaluatorBuilder().getStackFrameTransformer());
+      }
+      throw e;
     } catch (URISyntaxException e) {
       throw new PklException(e.getMessage(), e);
     }
   }
 
+  private static String renderCycle(Set<URI> cycle) {
+    var sb = new StringBuilder();
+    sb.append("┌─>");
+    var isFirst = true;
+    for (URI uri : cycle) {
+      if (isFirst) {
+        isFirst = false;
+      } else {
+        sb.append("\n│");
+      }
+      sb.append("\n│  ");
+      sb.append(uri.toString());
+    }
+    sb.append("\n└─ ");
+    return sb.toString();
+  }
+
+  private static Set<Set<URI>> findImportCycle(ModuleSource moduleSource) {
+    var builder = evaluatorBuilder();
+    var analyzer =
+        new Analyzer(
+            StackFrameTransformers.defaultTransformer,
+            SecurityManagers.defaultManager,
+            builder.getModuleKeyFactories(),
+            builder.getModuleCacheDir(),
+            builder.getProjectDependencies(),
+            builder.getHttpClient());
+    var graph = analyzer.importGraph(moduleSource.getUri());
+    var res = new HashSet<Set<URI>>();
+    for (var uri : graph.imports().keySet()) {
+      if ("pkl".equals(uri.getScheme())) {
+        continue;
+      }
+      var cycle = doFindCycle(uri, graph, new ArrayList<>(List.of(uri)));
+      if (cycle != null) {
+        res.add(new TreeSet<>(cycle));
```

Comment:
Had to use a `Set` here because otherwise we'd report on every permutation of a cycle.